pkg/dockerregistry/server/quotarestrictedblobstore.go

```diff
@@ -120,26 +161,103 @@ func (bw *quotaRestrictedBlobWriter) Commit(ctx context.Context, provisional dis
+	if err := admitLimitRanges(ctx, repo, size); err != nil {
+		return err
+	}
+
+	if err := admitQuotas(ctx, repo); err != nil {
 		return err
 	}
 
+	return nil
+}
+
+// admitLimitRanges checks the blob against any established limit ranges.
+func admitLimitRanges(ctx context.Context, repo *repository, size int64) error {
+	if size < 1 {
+		return nil
+	}
+
+	var (
+		lrs *kapi.LimitRangeList
+		err error
+	)
+
+	if projectLimitRanges != nil {
+		obj, exists, _ := projectLimitRanges.Get(repo.namespace)
+		if exists {
+			lrs = obj.(*kapi.LimitRangeList)
+		}
+	}
+	if lrs == nil {
+		context.GetLogger(ctx).Debugf("Listing limit ranges in namespace %s", repo.namespace)
+		lrs, err = repo.limitClient.LimitRanges(repo.namespace).List(kapi.ListOptions{})
+		if err != nil {
+			context.GetLogger(ctx).Errorf("Failed to list limitranges: %v", err)
+			return err
+		}
+		if projectLimitRanges != nil {
+			err = projectLimitRanges.Add(repo.namespace, lrs)
+			if err != nil {
+				context.GetLogger(ctx).Errorf("Failed to cache limit range list: %v", err)
+			}
+		}
+	}
+
+	for _, limitrange := range lrs.Items {
+		context.GetLogger(ctx).Debugf("Processing limit range %s/%s", limitrange.Namespace, limitrange.Name)
+		for _, limit := range limitrange.Spec.Limits {
+			if err := imageadmission.AdmitImage(size, limit); err != nil {
+				errMsg := fmt.Sprintf("Refusing to write blob exceeding limit range: %v", err)
+				context.GetLogger(ctx).Error(errMsg)
+				reportQuotaExceededEvent(ctx, repo, nil, "BlobWriteDenied", errMsg)
+				return distribution.ErrAccessDenied
+			}
+		}
+	}
+
+	return nil
+}
+
+// admitQuotas checks the blob against any established quotas.
+func admitQuotas(ctx context.Context, repo *repository) error {
+	var (
+		rqs *kapi.ResourceQuotaList
+		err error
+	)
+
+	if projectResourceQuotas != nil {
+		obj, exists, _ := projectResourceQuotas.Get(repo.namespace)
+		if exists {
+			rqs = obj.(*kapi.ResourceQuotaList)
+		}
+	}
+	if rqs == nil {
+		context.GetLogger(ctx).Debugf("Listing resource quotas in namespace %s", repo.namespace)
+		rqs, err = repo.quotaClient.ResourceQuotas(repo.namespace).List(kapi.ListOptions{})
+		if err != nil {
+			context.GetLogger(ctx).Errorf("Failed to list resourcequotas: %v", err)
+			return err
+		}
+		if projectResourceQuotas != nil {
+			err = projectResourceQuotas.Add(repo.namespace, rqs)
+			if err != nil {
+				context.GetLogger(ctx).Errorf("Failed to cache resource quota list: %v", err)
+			}
+		}
+	}
+
 	usage := kapi.ResourceList{
 		// we are about to tag a single image to an image stream
-		imageapi.ResourceImages: *resource.NewQuantity(1, resource.DecimalSI),
+		imageapi.ResourceImageStreamImages: *resource.NewQuantity(1, resource.DecimalSI),
 	}
 	resources := quota.ResourceNames(usage)
 
 	for _, rq := range rqs.Items {
+		context.GetLogger(ctx).Debugf("Processing resource quota %s/%s", rq.Namespace, rq.Name)
 		newUsage := quota.Add(usage, rq.Status.Used)
```

Comment:
> If so, seems like it might be comment worthy. It wasn't super-obvious that this it was safe to me

nm, I see now that these methods are creating copies, not adding to originals.